nomad/core_sched.go

```diff
@@ -902,40 +902,49 @@ func (c *CoreScheduler) rootKeyRotateOrGC(eval *structs.Evaluation) error {
 	// a rotation will be sent to the leader so our view of state
 	// is no longer valid. we ack this core job and will pick up
 	// the GC work on the next interval
-	wasRotated, err := c.rootKeyRotate(eval)
+	wasRotated, err := c.rootKeyRotate(eval, time.Now())
 	if err != nil {
 		return err
 	}
 	if wasRotated {
 		return nil
 	}
-	return c.rootKeyGC(eval)
+	return c.rootKeyGC(eval, time.Now())
 }
 
-func (c *CoreScheduler) rootKeyGC(eval *structs.Evaluation) error {
-
-	oldThreshold := c.getThreshold(eval, "root key",
-		"root_key_gc_threshold", c.srv.config.RootKeyGCThreshold)
+func (c *CoreScheduler) rootKeyGC(eval *structs.Evaluation, now time.Time) error {
 
 	ws := memdb.NewWatchSet()
 	iter, err := c.snap.RootKeyMetas(ws)
 	if err != nil {
 		return err
 	}
 
+	// the threshold is longer than we can support with the time table, and we
+	// never want to force-GC keys because that will orphan signed Workload
+	// Identities
+	rotationThreshold := now.Add(-1 *
+		(c.srv.config.RootKeyRotationThreshold + c.srv.config.RootKeyGCThreshold))
+
 	for {
 		raw := iter.Next()
 		if raw == nil {
 			break
 		}
 		keyMeta := raw.(*structs.RootKeyMeta)
-		if keyMeta.Active() || keyMeta.Rekeying() {
-			continue // never GC the active key or one we're rekeying
+		if !keyMeta.Inactive() {
+			continue // never GC keys we're still using
 		}
-		if keyMeta.CreateIndex > oldThreshold {
-			continue // don't GC recent keys
+
+		c.logger.Trace("checking inactive key eligibility for gc",
+			"create_time", keyMeta.CreateTime, "threshold", rotationThreshold.UnixNano())
+
+		if keyMeta.CreateTime > rotationThreshold.UnixNano() {
```

Comment:
When could this ever happen?

Reply:
If I rotate a key without prepublishing it'll be active and be used for signing. If I rotate it again, it'll be inactive and therefore eligible for GC, but it will be recent enough that there are still signed workload identities that have not expired. We don't keep a record in the state store of signed WI's (unlike Variables), so we need to make sure we're not deleting a key that's still in use.

Reply:
got it, thanks